launch/launch/utilities/type_utils.py

```diff
@@ -0,0 +1,565 @@
+        else:
+            assert len(member_types) == 1
+        assert len(member_types) in (1, 2)
+        valid_types = tuple(member_types)
+        member_types.discard(str)
+        try:
+            member_type = member_types.pop()
+        except KeyError:
+            member_type = str
+        return (
+            all(isinstance(x, valid_types) for x in value[1:]) and
+            member_type in ScalarTypesTuple
+        )
+    return isinstance(value, ScalarTypesTuple)
+
+
+def is_instance_of(
+    value: Any,
+    data_type: Optional[AllowedTypesType] = None,
+    can_be_str: bool = False,
+) -> bool:
+    """
+    Check if `value` is instance of `data_type`.
+
+    :param value: variable to check.
+    :param data_type: class that `value` should be instance of.
+        When `None` the function behaves like :py:func:`is_instance_of_valid_type`.
+    :param can_be_str: if `True`, strings will also be accepted.
+      :py:mod:`launch.frontend` makes use of this for string embedded substitutions.
+    :raise: `ValueError` if `data_type` is invalid.
+    :return: `True` if `value` is an instance of `data_type`, else `False`.
+    """
+    if data_type is None:
+        return is_instance_of_valid_type(value, can_be_str)
+    type_obj, is_list = extract_type(data_type)
+    if can_be_str:
+        type_obj = (str, type_obj)
+    if not is_list:
+        return isinstance(value, type_obj)
+    if not isinstance(value, list):
+        return False
+    return all(isinstance(x, type_obj) for x in value)
+
+
+def coerce_to_type(
+    value: Text,
+    data_type: Optional[AllowedTypesType] = None,
+    can_be_str: bool = False,
+) -> StrSomeValueType:
+    """
+    Coerce `value` type to `data_type`.
+
+    :param value: string to be coerced.
+    :param data_type: value will be coerced to data_type.
+    :param can_be_str: if `True`, the result will be kept as an string if it cannot be coerced.
+      In the case of lists, it will also accept strings as items.
+      :py:mod:`launch.frontend` makes use of this for string embedded substitutions.
+    :raise: `ValueError` if the coercion failed or `data_type` is invalid.
+    :raise: `TypeError` if `value` is not a `str`.
+    :return: `value` coerced to `data_type`.
+    """
+    def convert_as_yaml(value, error_msg):
+        try:
+            output = yaml.safe_load(value)
+        except Exception as err:
+            if can_be_str:
+                return value
+            raise ValueError(f'{error_msg}: yaml.safe_load() failed\n{err}')
+
+        if not is_instance_of_valid_type(output, can_be_str):
+            raise ValueError(
+                f'{error_msg}: output type is not allowed, got {type(output)}'
+            )
+        return output
+
+    ensure_argument_type(value, str, 'value', 'coerce_to_type')
+    if data_type is None:
+        # No type specified, return best conversion.
+        return convert_as_yaml(value, f"Failed to convert '{value}' using yaml rules")
```

Comment:
@ivanpauno meta, safe to ignore: looking at this function, it seems unconditionally converting to `yaml` and doing type checks if `data_type` is not `None` may be simpler.

Reply:
In some cases, the yaml conversion isn't used.
e.g.: for `int` or `float`

Reply:
Right, but it could be i.e. it'd have the same effect.